src/services/ontologyService.utils.ts

```diff
@@ -0,0 +1,98 @@
+    const promise = await getSchemaById(id);
+    const parsedOntology = await parseSingleOntology(promise);
+    return createOntologyObject(parsedOntology, relatedShapes);
+  });
+
+  return await Promise.all(promises);
+}
+
+export const parseSingleOntology = (item: string) => {
+  const parser = new N3.Parser();
+  const quads: Quad[] = [];
+  parser.parse(item,
+    (error, quad) => {
+      if (quad) {
+        quads.push(quad);
+      }
+    });
+  return quads;
+}
+
+export const createOntologyObject = (quads: Quad[], relatedShapes: Shape[]): Ontology => {
+  const firstSubject = quads.length > 0 ? quads[0].subject.id : 'No subject available!';
+  const nodes: { id: string; label: string; type: string }[] = [];
+  const links: { source: string; target: string }[] = [];
+
+  let subject = firstSubject;
+  let contributors: string[] = [];
+  let description = 'No description available!';
+  let version = 'No version available!';
+
+  // Create a map to keep track of the types of each subject
+  let typesMap: { [key: string]: string } = {};
+  quads.forEach(quad => {
+    const subjectId = quad.subject.id;
+    const predicateId = quad.predicate.id;
+    const objectId = quad.object.id;
+
+    // Track types of subjects
+    if (predicateId === 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type') {
+      typesMap[subjectId] = objectId;
+    }
+
+    if (predicateId === 'http://www.w3.org/2000/01/rdf-schema#label' || predicateId ==='http://www.w3.org/2002/07/owl#ObjectProperty') {
+      nodes.push({ id: subjectId, label: objectId, type: typesMap[subjectId] || 'Unknown' });
+    } else if (predicateId === 'http://www.w3.org/2000/01/rdf-schema#subClassOf' || predicateId ==='http://www.w3.org/2000/01/rdf-schema#domain' || predicateId ==='http://www.w3.org/2000/01/rdf-schema#range') {
+      links.push({ source: subjectId, target: objectId });
+    }
+
+    // get the ontology information
+    if (subjectId === firstSubject) {
+      switch (predicateId) {
+      case 'http://purl.org/dc/terms/contributor':
+        contributors.push(objectId.replace(/(^"|"$)/g, ''));
+        break;
+      case 'http://www.w3.org/2000/01/rdf-schema#label':
+        description = objectId.replace(/(^"|"$)/g, '').split('"@')[0];
+        break;
+      case 'http://www.w3.org/2002/07/owl#versionInfo':
+        version = objectId.replace(/(^"|"$)/g, '').split('"^^')[0];
+        break;
+      }
+    }
+  });
+
+  // Update nodes with type information
+  nodes.forEach(node => {
+    node.type = typesMap[node.id] || 'Unknown';
+  });
+
+  return {
+    subject,
+    contributors,
+    description,
+    version,
+    nodes,
+    links,
+    relatedShapes,
+  };
+};
+
+export const getOntologyById = async (id: string) => {
+  const relatedShapes = await getRelatedShapes(id);
+  const response = await getSchemaById(id);
+  const parsedOntology = await parseSingleOntology(response);
+  return createOntologyObject(parsedOntology, relatedShapes);
```

Comment:
I need the result from the previous line in each line. Is it therefore not actually necessary to wait or to execute the lines sequentially?

Reply:
I need the result from the previous line in each line. Is it therefore not actually necessary to wait or to execute the lines sequentially?

Reply:
The aswere is that is not a problem if you do not execute sequentially the two backend requests. You only have to make sure that both data has arrived when you start creating the ontolgy object.  The `Promise.all([relatedShapesPromise, parsedOntologyPromise])` makes sure of that. This code waits until both data is arrived. 

You might ask what's the difference then or is there any difference at all between the two. Yes, there is. In the original code 
```typescript
  const relatedShapes = await getRelatedShapes(id);
  const response = await getSchemaById(id);
````
A request to the backend will be sent in order to get the shapes and you wait until the answere arrives and only the start sending the second request to the backend. Of cours that will take time so the secod request has to wait until the answer for the first one arrives. For simplicity lets consider it take 300ms for both call to send a request and get an answere. Because the second request waits the first one it will only be sent after 300 ms and when the answere to the second call arrives in total it will be 600 ms

In this new vesion 
```typescript
   const relatedShapesPromise = getRelatedShapes(id);
    const parsedOntologyPromise = getSchemaById(id).then((response) => parseSingleOntology(response));
```
you send the two requests in almost the same time they do not have to wait each other. So the response for the  request probably will arrive at almost the same time very close to each other. In a concrete example that will mean that you send the first request and after 10ms probably you send the second one too. So after 300ms the response for the first arrives and after 10ms arrives the seocond aswere too. In total you have to wait 310 ms which is half of the previouse one.

It is a small optimization. The loading of the data took 1.5 seconds less. It still consumes alot of time because there are other probles too with this loading of the data. It loads the same data multiple times. I took already care of that in another pull request. So I did not mentioned here.